src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs

```diff
@@ -570,6 +631,107 @@ private static SyntaxList<TypeParameterConstraintClauseSyntax> GetConstraintClau
+
+            static void mergeConstraints(ref TypeParameterConstraintKind mergedKind, ImmutableArray<TypeWithAnnotations> originalConstraintTypes, TypeParameterConstraintClause clause)
+            {
+                if ((mergedKind & (TypeParameterConstraintKind.AllNonNullableKinds | TypeParameterConstraintKind.NotNull)) != (clause.Constraints & (TypeParameterConstraintKind.AllNonNullableKinds | TypeParameterConstraintKind.NotNull)))
+                {
+                    mergedKind |= TypeParameterConstraintKind.PartialMismatch;
+                }
+
+                if ((mergedKind & TypeParameterConstraintKind.ReferenceType) != 0 && (clause.Constraints & TypeParameterConstraintKind.ReferenceType) != 0)
+                {
+                    // Try merging nullability of a 'class' constraint
+                    TypeParameterConstraintKind clause1Constraints = mergedKind & TypeParameterConstraintKind.AllReferenceTypeKinds;
+                    TypeParameterConstraintKind clause2Constraints = clause.Constraints & TypeParameterConstraintKind.AllReferenceTypeKinds;
+                    if (clause1Constraints != clause2Constraints)
+                    {
+                        if (clause1Constraints == TypeParameterConstraintKind.ReferenceType) // Oblivious
+                        {
+                            // Take nullability from clause2
+                            mergedKind = (mergedKind & (~TypeParameterConstraintKind.AllReferenceTypeKinds)) | clause2Constraints;
+                        }
+                        else if (clause2Constraints != TypeParameterConstraintKind.ReferenceType)
+                        {
+                            // Neither nullability is oblivious and they do not match. Cannot merge.
+                            mergedKind |= TypeParameterConstraintKind.PartialMismatch;
+                        }
+                    }
+                }
+
+                if (originalConstraintTypes.Length == 0 && clause.ConstraintTypes.Length == 0)
+                {
+                    // Try merging nullability of implied 'object' constraint
+                    if (((mergedKind | clause.Constraints) & ~(TypeParameterConstraintKind.ObliviousNullabilityIfReferenceType | TypeParameterConstraintKind.Constructor)) == 0 &&
+                        (mergedKind & TypeParameterConstraintKind.ObliviousNullabilityIfReferenceType) != 0 && // 'object~'
+                        (clause.Constraints & TypeParameterConstraintKind.ObliviousNullabilityIfReferenceType) == 0)   // 'object?' 
+                    {
+                        // Merged value is 'object?'
+                        mergedKind &= ~TypeParameterConstraintKind.ObliviousNullabilityIfReferenceType;
+                    }
+                }
+            }
+        }
+
+        private ImmutableArray<TypeParameterConstraintClause> AdjustConstraintKindsBasedOnConstraintTypes(ImmutableArray<TypeParameterConstraintClause> constraintClauses)
```

Comment:
>AdjustConstraintKindsBasedOnConstraintTypes [](start = 62, length = 43)

nit: consider inlining (only one call site)

Reply:
> (only one call site)

I don't think this is the case

---
In reply to: [520169876](https://github.com/dotnet/roslyn/pull/49227#discussion_r520169876) [](ancestors = 520169876)

Reply:
This is a private method (which was introduced in this PR) and the only call site in this file that I can see is that line 447. Is there another?

---
In reply to: [520171684](https://github.com/dotnet/roslyn/pull/49227#discussion_r520171684) [](ancestors = 520171684,520169876)